File: src/main/并查集/t684/FindRedundantConnection.java
package 并查集.t684;

/**
 * @author: sotowang
 * @date: 2020/09/17 16:26
 */
public class FindRedundantConnection {
    public int[] findRedundantConnection(int[][] edges) {
        for(int i=edges.length-1;i>=0;i--){
            if(!isCircle(edges,i)){
                return edges[i];
            }
        }
        return new int[]{};
    }

    private boolean isCircle(int[][] edges,int removeIndex){
        UnionFind uf = new UnionFind(edges.length+1);
        for(int i=0;i<edges.length;i++){
            if(i == removeIndex){
                continue;
            }
            //若两点在一个并查集中，则构成了环
            if(!uf.union(edges[i][0],edges[i][1])){
                return true;
            }
        }
        return false;
    }
    class UnionFind{
        int[] parent;
        public UnionFind(int n){
            this.parent = new int[n];
            for(int i=0;i<n;i++){
                parent[i] = i;
            }
        }
        public boolean union(int x,int y){
            int rootX = find(x);
            int rootY = find(y);
            if(rootX == rootY){
                return false;
            }
            parent[rootX] = rootY;
            return true;
        }
        public int find(int x){
            while(parent[x] != x){
                x = parent[parent[x]];
            }
            return x;
        }
    }
}

File: src/main/并查集/t684/findRedundantConnection.go
package t684

type unionFind struct {
	parent []int
}

func generate(n int) *unionFind {
	uf := unionFind{}
	uf.parent = make([]int, n)
	for i := 0; i < n; i++ {
		uf.parent[i] = i
	}
	return &uf
}
func (uf *unionFind) union(x, y int) bool {
	rootX, rootY := uf.find(x), uf.find(y)
	if rootX == rootY {
		return false
	}
	uf.parent[rootX] = rootY
	return true
}
func (uf *unionFind) find(x int) int {
	for uf.parent[x] != x {
		x = uf.parent[uf.parent[x]]
	}
	return x
}

func isCircle(edges [][]int, removeIndex int) bool {
	uf := generate(len(edges) + 1)
	for i := range edges {
		if i == removeIndex {
			continue
		}
		if !uf.union(edges[i][0], edges[i][1]) {
			return true
		}
	}
	return false

}
func findRedundantConnection(edges [][]int) []int {
	for i := len(edges) - 1; i >= 0; i-- {
		if !isCircle(edges, i) {
			return edges[i]
		}
	}
	return []int{}
}
